libkmod/libkmod-file.c

```diff
@@ -58,57 +58,54 @@ static const struct comp_type {
 	// clang-format on
 };
 
-struct kmod_elf *kmod_file_get_elf(struct kmod_file *file)
+int kmod_file_get_elf(struct kmod_file *file, struct kmod_elf **elf)
 {
-	int err;
-
-	if (file->elf)
-		return file->elf;
-
-	err = kmod_file_load_contents(file);
-	if (err) {
-		errno = -err;
-		return NULL;
+	if (!file->elf) {
+		int err = kmod_file_load_contents(file);
+		if (err)
+			return err;
+
+		err = kmod_elf_new(file->memory, file->size, &file->elf);
+		if (err)
+			return err;
 	}
 
-	err = kmod_elf_new(file->memory, file->size, &file->elf);
-	if (err) {
-		errno = -err;
-		return NULL;
-	}
-	return file->elf;
+	*elf = file->elf;
+	return 0;
 }
 
-struct kmod_file *kmod_file_open(const struct kmod_ctx *ctx, const char *filename)
+int kmod_file_open(const struct kmod_ctx *ctx, const char *filename,
+		   struct kmod_file **out_file)
 {
 	struct kmod_file *file;
 	char buf[7];
 	ssize_t sz;
+	int ret;
 
 	assert_cc(sizeof(magic_zstd) < sizeof(buf));
 	assert_cc(sizeof(magic_xz) < sizeof(buf));
 	assert_cc(sizeof(magic_zlib) < sizeof(buf));
 
 	file = calloc(1, sizeof(struct kmod_file));
 	if (file == NULL)
-		return NULL;
+		return -ENOMEM;
 
 	file->fd = open(filename, O_RDONLY | O_CLOEXEC);
 	if (file->fd < 0) {
 		free(file);
-		return NULL;
+		return -ENOMEM;
```

Comment:
ENOMEM here doesn't feel right. However using the errno from open() after the free() is also invalid. Options:  a) add a local saved_errno in the error path so you can return it; b) annotate file with _cleanup_free_, and assign it to NULL after setting *out_file.  Approach (b) would allow to remove one more free() call in the error path.

Reply:
> Not sure I follow - why is it invalid? The free() call should never alter errno... does it?

humn.. true

> Either way, option b) applied. Thanks for the review.

thanks